refactor(TEL-2225): Update client actuator to 2.1.0

FranciscoMota opened the pull request:
This required a complete refactor of the code as of spring 2.0
actuator relies on Micrometer to collect and export metrics

Commit:
refactor(TEL-2225): Update client actuator to 2.1.0

This required a complete refactor of the code as of spring 2.0
actuator relies on Micrometer to collect and export metrics

Commit:
refactor(TEL-2225): Update readme

FranciscoMota commented on src/main/java/com/statful/client/framework/springboot/registry/StatfulMetricRegistry.java:
```diff
@@ -0,0 +1,190 @@
+
+        long value = (new Double(histogramSnapshot.mean(this.getBaseTimeUnit()))).longValue();
+
+        this.statfulClient.timer(meterId.getName(), value).with().tags(tags).send();
```
@senzzzi I had some thoughts regarding this one. Since it already applies a mean I thought it might make sense to call aggregatedTimer() instead but since you were already using it like this I decided to leave it.

gabrielmpinto commented on src/main/java/com/statful/client/framework/springboot/registry/StatfulMetricRegistry.java:
```diff
@@ -0,0 +1,190 @@
+
+        applyCustomFilter();
+
+        start();
```
This is deprecated, should be receiving a ThreadFactory. Maybe do `start(Executors.defaultThreadFactory());`for this constructor, and add a new constructor that receives the needed ThreadFactory.

gabrielmpinto commented on src/main/java/com/statful/client/framework/springboot/registry/StatfulMetricRegistry.java:
```diff
@@ -0,0 +1,190 @@
+    private void applyCustomFilter() {
+        this.config()
+                .meterFilter(
+                        MeterFilter.denyUnless(id ->
```
these MeterFilters could have their own methods to make this more readable

gabrielmpinto commented on src/main/java/com/statful/client/framework/springboot/registry/StatfulMetricRegistry.java:
```diff
@@ -0,0 +1,190 @@
+                .map(tag -> {
+                    String[] splitTag = tag.split("=");
+
+                    return new Tag() {
```
Why are you creating an anonymous class instead of using `Tag.of(splitTag[0], splitTag[1])`?

FranciscoMota replied:
You're right, much simpler that way. Thanks :)

gabrielmpinto commented on src/main/java/com/statful/client/framework/springboot/registry/StatfulMetricRegistry.java:
```diff
@@ -0,0 +1,190 @@
+                    this.writeCounter((Counter) meter);
+                } else if (meter instanceof FunctionCounter) {
+                    this.writeCounter((FunctionCounter) meter);
+                } else if (meter instanceof LongTaskTimer) {
```
Empty if

gabrielmpinto commented on src/main/java/com/statful/client/framework/springboot/registry/StatfulMetricRegistry.java:
```diff
@@ -0,0 +1,190 @@
+
+        meterId.getTags().forEach((tag) -> tags.putTag(tag.getKey(), tag.getValue()));
+
+        this.statfulClient.counter(meterId.getName(), (new Double(counter.count())).intValue()).with().tags(tags).send();
```
Please use Double.valueOf() instead of the deprecated Double constructor

gabrielmpinto commented on src/main/java/com/statful/client/framework/springboot/registry/StatfulMetricRegistry.java:
```diff
@@ -0,0 +1,190 @@
+
+        meterId.getTags().forEach((tag) -> tags.putTag(tag.getKey(), tag.getValue()));
+
+        long value = (new Double(histogramSnapshot.mean(this.getBaseTimeUnit()))).longValue();
```
Please use Double.valueOf() instead of the deprecated Double constructor

gabrielmpinto commented on src/main/java/com/statful/client/framework/springboot/registry/StatfulRegistryAutoConfiguration.java:
```diff
@@ -0,0 +1,26 @@
+
+@Configuration
+@EnableConfigurationProperties(value = StatfulMetricsProperties.class)
+public class StatfulRegistryAutoConfiguration {
```
This should be in a different package (different maven module IMO, see other comment)

gabrielmpinto reviewed (commented):
I'd rather have StatfulMetricRegistry as a separate package, and dependency. That way I could use Statful's micrometer registry without depending on Spring Boot configurations. A separate statful-micrometer-registry would be more appropriate I believe. And this one would autoconfigure that repo for Spring Boot

Commit:
refactor(TEL-2225): Extract statful metric registry to external project

FranciscoMota commented:
Refactored this as suggested in feedback. Metric Registry was moved to https://github.com/statful/statful-micrometer-registry/

FranciscoMota committed:
chore(TEL-2225): add java client configuration, including @Timer aspect

gabrielmpinto commented:
Please make sure the travis build passes

FranciscoMota committed:
chore(TEL-2225): add missing configurations

This makes it so it's only required to add the dependency and set the token to start using the client

FranciscoMota committed:
fix(TEL-2225): allow metric collection by adding dependency and defining API token

FranciscoMota committed:
fix(TEL-2225): enable micrometer

FranciscoMota committed:
fix(TEL-2225): add annotation that initializes StatfulClient only when it's missing
